File: typescript/src/adapters/openai.ts
import OpenAI from 'openai';
import { BaseAdapter, GenerateOptions } from './base.js';

/**
 * Adapter for OpenAI's chat models (GPT-4o, GPT-4.1, etc.).
 *
 * Uses the `openai` SDK client. Structured output is achieved via
 * OpenAI's `response_format` with `json_schema` mode, which guarantees
 * the model produces JSON conforming to the provided schema.
 */
export class OpenAIAdapter extends BaseAdapter {
  private client: OpenAI;
  private _model: string;
  private defaultOptions: GenerateOptions;

  /**
   * @param model - OpenAI model ID (default `'gpt-4o'`).
   * @param apiKey - OpenAI API key. Falls back to `OPENAI_API_KEY` env var.
   * @param options - Default generation options applied to every call.
   */
  constructor(
    model: string = 'gpt-4o',
    apiKey?: string,
    options?: GenerateOptions,
  ) {
    super();
    this._model = model;
    this.client = new OpenAI({ apiKey: apiKey ?? process.env['OPENAI_API_KEY'] });
    this.defaultOptions = options ?? {};
  }

  /** The model identifier used for API calls. */
  get modelName(): string {
    return this._model;
  }

  /**
   * Generate free-form text.
   *
   * @param systemPrompt - System-level instruction.
   * @param userPrompt - User message.
   * @param options - Per-call overrides merged over constructor defaults.
   * @returns The text content of the assistant's response.
   */
  async generate(
    systemPrompt: string,
    userPrompt: string,
    options?: GenerateOptions,
  ): Promise<string> {
    const merged = { ...this.defaultOptions, ...options };
    const response = await this.client.chat.completions.create({
      model: this._model,
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: userPrompt },
      ],
      max_tokens: merged.maxTokens,
      temperature: merged.temperature,
    });

    return response.choices[0]?.message?.content ?? '';
  }

  /**
   * Generate structured (typed) output using OpenAI's `json_schema` response format.
   *
   * The model is constrained to produce a JSON object matching the provided schema.
   * The result is parsed with a simple `JSON.parse` helper.
   *
   * @typeParam T - The expected output type.
   * @param systemPrompt - System-level instruction.
   * @param userPrompt - User message.
   * @param outputSchema - JSON Schema object describing the desired output shape.
   * @param options - Per-call overrides.
   * @returns A parsed object of type T.
   */
  async generateStructured<T>(
    systemPrompt: string,
    userPrompt: string,
    outputSchema: Record<string, unknown>,
    options?: GenerateOptions,
  ): Promise<T> {
    const merged = { ...this.defaultOptions, ...options };

    const response = await this.client.chat.completions.create({
      model: this._model,
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: userPrompt },
      ],
      max_tokens: merged.maxTokens,
      temperature: merged.temperature,
      response_format: {
        type: 'json_schema',
        json_schema: {
          name: 'output',
          strict: true,
          schema: outputSchema as Record<string, unknown>,
        },
      },
    });

    const raw = response.choices[0]?.message?.content ?? '{}';
    return this.parse<T>(raw);
  }

  /**
   * Safely parse a JSON string into type T.
   * Logs a warning and returns an empty object on parse failure.
   */
  private parse<T>(raw: string): T {
    try {
      return JSON.parse(raw) as T;
    } catch {
      console.warn('OpenAIAdapter: failed to parse structured output, returning empty object');
      return {} as T;
    }
  }
}

File: python/src/dual_agent_sdk/adapters/openai.py
"""OpenAI adapter.

Provides :class:`OpenAIAdapter`, which calls the OpenAI Chat Completions
API via the official ``openai`` SDK.  Structured generation leverages the
``response_format`` parameter with ``json_schema``.
"""

from __future__ import annotations

import json
import os
import re
from typing import Any

from openai import AsyncOpenAI

from .base import BaseAdapter


class OpenAIAdapter(BaseAdapter):
    """Adapter for OpenAI's chat models.

    Parameters:
        model: OpenAI model ID (default ``"gpt-4o"``).
        api_key: OpenAI API key.  If ``None`` the adapter reads the
            ``OPENAI_API_KEY`` environment variable.
        **kwargs: Extra keyword arguments forwarded to
            :class:`openai.AsyncOpenAI` (e.g. ``base_url``).
    """

    def __init__(
        self,
        model: str = "gpt-4o",
        api_key: str | None = None,
        **kwargs: Any,
    ) -> None:
        resolved_key = api_key or os.environ.get("OPENAI_API_KEY")
        if not resolved_key:
            raise ValueError(
                "OpenAI API key must be provided or set via "
                "the OPENAI_API_KEY environment variable"
            )
        self._model = model
        self._client = AsyncOpenAI(api_key=resolved_key, **kwargs)

    # ------------------------------------------------------------------
    # BaseAdapter interface
    # ------------------------------------------------------------------

    async def generate(
        self, system_prompt: str, user_prompt: str, **kwargs: Any
    ) -> str:
        """Generate a text completion via the Chat Completions API."""
        response = await self._client.chat.completions.create(
            model=self._model,
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt},
            ],
            max_tokens=kwargs.pop("max_tokens", 4096),
            **kwargs,
        )
        content = response.choices[0].message.content
        return content or ""

    async def generate_structured(
        self,
        system_prompt: str,
        user_prompt: str,
        output_schema: dict[str, Any],
        **kwargs: Any,
    ) -> dict[str, Any]:
        """Generate a structured response using ``response_format``.

        Uses the ``json_schema`` response format available in recent
        OpenAI models.  Falls back to JSON extraction from the text
        response if the API does not support this mode.
        """
        schema_name = kwargs.pop("schema_name", "output")

        try:
            response = await self._client.chat.completions.create(
                model=self._model,
                messages=[
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_prompt},
                ],
                max_tokens=kwargs.pop("max_tokens", 4096),
                response_format={
                    "type": "json_schema",
                    "json_schema": {
                        "name": schema_name,
                        "schema": output_schema,
                        "strict": True,
                    },
                },
                **kwargs,
            )
            content = response.choices[0].message.content
            if content is None:
                raise RuntimeError("OpenAI returned empty content")
            return json.loads(content)
        except Exception:
            # Fallback: request text and extract JSON from the response.
            text = await self.generate(
                system_prompt=system_prompt,
                user_prompt=user_prompt
                + "\n\nRespond ONLY with a JSON object matching the schema. Do not wrap in markdown fences.",
                **kwargs,
            )
            return self._extract_json(text)

    @staticmethod
    def _extract_json(text: str) -> dict[str, Any]:
        """Best-effort JSON extraction from a text block.

        Strips optional markdown code fences and parses the first JSON
        object found.
        """
        # Remove markdown fences if present.
        cleaned = re.sub(r"^```(?:json)?\s*", "", text.strip())
        cleaned = re.sub(r"\s*```$", "", cleaned)

        # Try a direct parse first.
        try:
            return json.loads(cleaned)
        except json.JSONDecodeError:
            pass

        # Attempt to find a JSON object via brace matching.
        match = re.search(r"\{.*\}", cleaned, re.DOTALL)
        if match:
            try:
                return json.loads(match.group(0))
            except json.JSONDecodeError:
                pass

        raise ValueError(f"Could not extract valid JSON from response: {text[:500]}")

    @property
    def model_name(self) -> str:
        return self._model
